pkg/scheduler/plugins/reservation/hook.go

```diff
@@ -0,0 +1,209 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reservation
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+	resourceapi "k8s.io/kubernetes/pkg/api/v1/resource"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/frameworkext"
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+var (
+	_ frameworkext.PreFilterPhaseHook = &Hook{}
+	_ frameworkext.FilterPhaseHook    = &Hook{}
+)
+
+type Hook struct{}
+
+func NewHook() *Hook {
+	return &Hook{}
+}
+
+func (h *Hook) Name() string { return Name }
+
+func (h *Hook) PreFilterHook(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod) (*corev1.Pod, bool) {
+	// do not hook if the pod is a reserve pod
+	if IsReservePod(pod) {
+		return nil, false
+	}
+
+	// list reservations and nodes, and check each available reservation whether it matches the pod or not
+	matchedCache, err := prepareMatchReservationState(handle, pod)
+	if err != nil {
+		klog.Warningf("PreFilterHook failed to get matched reservations, err: %v", err)
+		return nil, false
+	}
+
+	// if the pod match any reservations, initialize a state for matched reservations.
+	// it stores the matched reservations meta for the pod, mapping from nodeName to RInfo
+	cycleState.Write(preFilterStateKey, matchedCache)
+	klog.V(4).InfoS("PreFilterHook successfully update reservation state for pod", "pod", klog.KObj(pod))
+
+	// skip pod pre-filter of affinities/anti-affinities, topology constrains
+	return preparePreFilterPod(pod), true
+}
+
+func (h *Hook) FilterHook(handle frameworkext.ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool) {
+	// do not hook if not reserve state (where we should check if pod match any reservation on node)
+	if IsReservePod(pod) {
+		return nil, nil, false
+	}
+	// abort the hook if node is not found
+	node := nodeInfo.Node()
+	if node == nil {
+		return nil, nil, false
+	}
+	// expected matchedCache is prepared in PreFilterHook
+	matchedCache := getPreFilterState(cycleState)
+	if matchedCache == nil {
+		return nil, nil, false
+	}
+	// skip hook when no reservation matched
+	rOnNode := matchedCache.GetOnNode(node.Name)
+	if len(rOnNode) <= 0 {
+		return nil, nil, false
+	}
+
+	// fix-up reserved resources and ports
+	return pod, prepareFilterNodeInfo(pod, nodeInfo, rOnNode), true
+}
+
+func prepareMatchReservationState(handle frameworkext.ExtendedHandle, pod *corev1.Pod) (*AvailableCache, error) {
+	rList, err := handle.KoordinatorSharedInformerFactory().Scheduling().V1alpha1().Reservations().Lister().List(labels.Everything())
```

Comment:
The cost of this function is quite large. Maybe we can consider some optimizations. 
For example, we establish the index relationship between NodeName and Reservation through the indexer mechanism provided by the shared informer for the Reservation that has been successfully scheduled. Then query whether the Node has Reservations through indexer. This avoids rebuilding newAvailableCache. In addition, we can also add an atomic counter to count how many reservation instances have been scheduled in the current cluster when creating an index. If this counter is 0, return directly.

Reply:
fixed. use index with `status.nodeName` here